### PR TITLE
EVG-17275 check parser project collection if patched parser project is empty

### PR DESCRIPTION
**Pull request by ablack12**
[EVG-17275](https://jira.mongodb.org/browse/EVG-17275)

### Description 
Stop setting patched parser project after a patch is finalized, since we store the thing in the parser project collection anyway.

### Testing 
Tested by [creating a patch](https://evergreen-staging.corp.mongodb.com/rest/v2/patches/636c137997b1d36b5c582ee3), verifying it could be scheduled / reconfigured / run. 

This is a bit risky since random operations do seem to depend on patched parser project. I audited every instance I could, First deploying the change that checks for version, _then_ the one that unsets the parser project, then we can deploy the unset parser project line, so that if we need to revert the code will know how to handle patches with an empty string.

### CODE REVIEW

**Comment by hadjri** [comment]
Now that we only need `PatchedParserProject` to make it to the patch finalization stage, can we update `FinalizePatch` to unset the field at the same time we add to the parser projects collection?

File: service/rest_patch.go [hunk 1]
```diff
@@ -62,6 +64,20 @@ func (restapi restAPI) getPatchConfig(w http.ResponseWriter, r *http.Request) {
 	}
 	w.Header().Set("Content-Type", "application/x-yaml; charset=utf-8")
 	w.WriteHeader(http.StatusOK)
-	_, err := w.Write([]byte(projCtx.Patch.PatchedParserProject))
+	var err error
+	if projCtx.Patch.PatchedParserProject != "" {
+		_, err = w.Write([]byte(projCtx.Patch.PatchedParserProject))
+	} else {
+		var pp *model.ParserProject
```
Comment (hadjri) [inline service/rest_patch.go:71]: This line can be removed

File: service/rest_patch.go [hunk 2]
```diff
@@ -62,6 +64,20 @@ func (restapi restAPI) getPatchConfig(w http.ResponseWriter, r *http.Request) {
...
+		pp, err = model.ParserProjectFindOneById(projCtx.Patch.Version)
+		if pp != nil {
+			var projBytes []byte
```
Comment (hadjri) [inline service/rest_patch.go:74]: Same comment

**Comment by hadjri** [comment]
I also thing `PatchedProjectConfig` also follows this pattern and can be removed after patch finalization. Can we create a ticket for that as well?

**Review by hadjri — CHANGES_REQUESTED** [review]

File: model/patch_lifecycle.go [hunk 3]
```diff
@@ -116,8 +117,9 @@ type PatchUpdate struct {
 // Returns an http status code and error.
 func ConfigurePatch(ctx context.Context, p *patch.Patch, version *Version, proj *ProjectRef, patchUpdateReq PatchUpdate) (int, error) {
 	var err error
-	project := &Project{}
-	if _, err := LoadProjectInto(ctx, []byte(p.PatchedParserProject), nil, p.Project, project); err != nil {
+	project, _, err := FindAndTranslateProjectForPatch(ctx, p)
+	if err != nil {
+		grip.Debug("issue here")
```
[the remark this answers is not in the document]
Comment (ablack12) [reply model/patch_lifecycle.go]: 🤦‍♀️

File: service/rest_patch.go (hunk 1 above)
Comment (ablack12) [reply service/rest_patch.go]: We need it to then return the error correctly at 82 (although i agree it's not clear so i'll restructure)

**Comment by ablack12** [comment]
Ah I think this bit from the description answers your first question:
> This is a bit risky since random operations do seem to depend on patched parser project. I audited every instance I could, First deploying the change that checks for version, then the one that unsets the parser project, then we can deploy the unset parser project line, so that if we need to revert the code will know how to handle patches with an empty string.

I forgot about patched config also having a collection. I'll add that into this.

**Review by hadjri — APPROVED** [review]
LGTM!

**Comment by ablack12** [comment]
Added a good amount of stuff if you can re-review @hadjri  (ended up being some refactors just from reading patched config code; it does look like generally this will already work though since find looks it up from the version): https://github.com/evergreen-ci/evergreen/blob/776ed1d14d3ca0835f20fe7d5ffaee38047f4300/model/project_ref.go#L533

File: model/patch_lifecycle.go [hunk 4]
```diff
@@ -649,7 +649,7 @@ func FinalizePatch(ctx context.Context, p *patch.Patch, requester string, github
 
 // getFullPatchParams will retrieve a merged list of parameters defined on the patch alias (if any)
 // with the parameters that were explicitly user-specified, with the latter taking precedence.
-func getFullPatchParams(p *patch.Patch) ([]patch.Parameter, error) {
+func getFullPatchParams(p *patch.Patch, config *ProjectConfig) ([]patch.Parameter, error) {
```
Comment (hadjri) [inline model/patch_lifecycle.go:663]: config seems to be unused?

File: units/patch_intent.go [hunk 5]
```diff
@@ -931,10 +931,8 @@ func (j *patchIntentProcessor) verifyValidAlias(projectId string, patchDoc *patc
 	if err != nil {
 		return errors.Wrapf(err, "retrieving aliases for project '%s'", projectId)
 	}
-	for _, a := range aliases {
-		if a.Alias == alias {
-			return nil
-		}
+	if len(aliases) > 0 {
+
```
Comment (hadjri) [inline units/patch_intent.go:935]: This block is empty

File: units/patch_intent.go [hunk 6]
```diff
@@ -914,15 +914,15 @@ func (j *patchIntentProcessor) buildTriggerPatchDoc(patchDoc *patch.Patch) error
...
-func (j *patchIntentProcessor) verifyValidAlias(projectId string, patchDoc *patch.Patch) error {
+func (j *patchIntentProcessor) verifyValidAlias(projectId string, configStr string) error {
 	alias := j.intent.GetAlias()
 	if alias == "" {
 		return nil
 	}
 	var projectConfig *model.ProjectConfig
-	if patchDoc.PatchedProjectConfig != "" {
+	if configStr != "" {
 		var err error
```
Comment (hadjri) [inline units/patch_intent.go:924]: This line can be removed

File: model/patch_lifecycle.go (hunk 4 above)
Comment (ablack12) [reply model/patch_lifecycle.go:663]: ah thank you; i was considering a refactor but decided against it

File: units/patch_intent.go (hunk 6 above)
Comment (ablack12) [reply units/patch_intent.go:924]: It can't be because projectConfig needs to be used outside of the if statement

**Review by hadjri — APPROVED** [review]

**Pull request merged**